강원대 BE_최현준 2단계 - 회원 로그인

Hyeonjun0527 opened the pull request:
## 주요 변경 사항

1.
멤버의 세컨더리 어댑터 계층인 퍼시스턴스 계층 구현
멤버의 어플리케이션 계층의 비지니스 흐름을 담당하는 `UseCase` 인터페이스 구현 및 그 인터페이스의 .구현체인 `MemberInteractor` 구현
회원가입 때 비밀번호를 해시화 하기 위해 `PasswordEncoder` 인터페이스와 객체 구현
2.
`JwtTokenProvider`,`JwtTokenPort`, `JwtAuthenticationFilter`,`AuthorizationAspect` 추가
토큰 유틸인 `JwtTokenProvider` 
`JwtTokenPort는` 토큰 유틸의 인터페이스
`JwtAuthenticationFilter는` 서블릿필터에서 1차권한검사
`AuthorizationAspect는` 관리자용 api에 접근했을 때 검사, `WebConfig는` `aspect` 추가 설정, `RequireAdmin` 커스텀 어노테이션으로 관리자용 api 지정
3.
`MemberController` 추가 
`AdminController` 추가 관리자 CRUD 기능

4.
`ErrorCode에` 401과 403 추가 `GlobalExceptionHandler에서` 처리
`ForbiddenException`, `UnauthorizedException` 에러 추가.
깃 관련 문제로 step1에서 `ErrorResponse를`  리턴하는게 아니라 `ProblemDetail를` 리턴하는 방식으로 리팩토링했었던 것을 step2에서 누락했음. 이를 `GlobalExceptionHandler에` 반영.
ErrorResponse응답방식과 ProblemDetail응답방식이 다르고, ProblemDetail이 
RFC 7807 표준 (웹 API 에러 응답의 국제 표준)을 준수하기 때문에 변경함
```
errorResponse방식
{
  "timestamp": "2025-01-09T14:15:16.123Z",
  "status": 400,
  "error": "Bad Request", 
  "message": "입력하신 정보를 다시 확인해주세요",
  "code": "VALIDATION_FAILED",
  "path": "/api/products"
}

problemDetail방식
{
  "type": "about:blank",
  "title": "Bad Request",
  "status": 400,
  "detail": "입력하신 정보를 다시 확인해주세요",
  "instance": "/api/products", 
  "code": "VALIDATION_FAILED",
  "path": "/api/products"
}

```
## 맞닥뜨린 문제 및 한계

스프링 시큐리티를 안쓰기로 결정했고, 그래서 패스워드 인코더를 직접 만들어야했다. 
`PasswordEncoder` 클래스를 처음에 만들었는데, `MemberInteractor` 패스워드 인코더의 구체적인 구현체를 모르게 하고 싶었다. 그래서 `PasswordEncoder` 인터페이스를 만들고 그 인터페이스를 구현한 구현체 `SimplePasswordEncoder를` 만들었다.

`JwtTokenProvider는` jwt 기반 인증/인가 시스템을 구현하기 위한 유틸리티이다.
토큰 생성, 검증, 정보추출을 처리한다.

테스트 코드에서 `Error creating bean with name 'jwtAuthenticationFilter` 문제 발생
테스트 코드에서 `@WebMvcTest`는 웹 계층만 로드하기 때문에 컨트롤러가 생성자·필드로 주입받는 일반 `@Component`, `@Service`‑계층 빈은 스캔 대상에서 제외된다.
그래서 `JwtAuthenticationFilter가` 생성자 주입으로 `JwtTokenPort를` 요구하는데 `Context에` `Bean이` 없어서 에러가 발생한다. 그래서 스프링 프로필에 따라 빈이 등록 여부를 컨트롤했다.
그리고 `MemberInteractor가` `JwtTokenPort를` 의존성으로 필요로 하여 `MockJwtTokenProvider가` 필요했다... 다른 해결책으로 `@TestConfiguration`을 사용하는 방법과 테스트코드에 `MockitoBean`으로 등록하는 것이 있었는데 첫번째 방법은 매번 import해야하고 두번쨰 방법은 매번 `MockitoBean`을 추가해야하여 불편했다.


## 중점적으로 리뷰 받고 싶은 곳
JwtAuthenticationFilter으로 관리자용 api를 보호
JwtTokenProvider로 토큰 생성/검증/추출 하는 부분
컨트롤러가 각 모든 UseCase를 의존하는데 파사드 패턴으로 모든 Product 관련 UseCase를 묶어 Facade를 만들지 고민입니다. 그리고 UseCase라는 인터페이스명을 Port로 변경할 지 고민입니다.
## 추가 내용, 추후 추가 가능 내용
아랫 부분은 먼저 step3 위시리스트를 추가하고 적용할 수도 있을 내용입니다

Auth도메인 생성하고, 회원도메인과 분리하고 싶다...
(선택) 회원을 조회, 추가, 수정, 삭제할 수 있는 관리자 화면을 구현...
회원용 로그인 화면 구현
refresh token 도입하여 자동로그인 및 갱신



## 2차 리뷰
### JwtAuthenticationFilter에서 빠뜨렸던 Path 추가
인증을 거쳐서는 안되는 url을 누락했던 부분을 추가함.
### 테스트 환경 bean 분리
  - TestConfig, MockJwtTokenProvider
  MockJwtTokenProvider는 MemberInteractor에서 사용중이기에 Mock가 필요했음.
  - jwt.enabled뿐만 아니라 admin.enabled 으로 관리자 기능 on/off
- ProductTest, ProductControllerTest에서 @ActiveProfiles 제거ConditionalOnProperty을 통해 jwt.enabled 값으로 빈 주입할지를on/off
- WebConfig와 AuthorizationAspect 등에서 혼란을 야기하는 profile 기반 설정 +  ConditionalOnProperty에서 오직 ConditionalOnProperty로만 관리

### order 제거
  - 현재 상황에 불필요한 @Order 제거

### 코드리팩토링 가독성 향상

- JwtTokenProvider에서 isValid 메서드를 isNotValid로 변경
- JwtAuthenticationFilter 에서 if문을 짧게 끊어 가독성 향상

### 테스트 코드 놓쳤던 것 다시 작동하게 변경!

### Usecase
모든 유스케이스를 묶은 하나의 유스케이스를 생성함(파사드 패턴)
기존에 클라이언트는 자신이 사용하지 않는 메서드에 의존하도록 강요받아서는 안된다. 인터페이스 분리원칙, 그리고 하나의 역할만 수행하라 단일책임원칙을 지키려고 하여 각각의 유스케이스를 분리했었는데, 
테스트 용이성과 보일러플레이트 감소를 위해 이걸 또 다시 통합했습니다.

Hyeonjun0527 committed:
refactor(product): 유스케이스 계층 도입 및 아키텍처 개선

Product 관련 기능 유지보수성, 테스트 용이성, 코드 명확성 향상을 위해 대규모 리팩토링 진행

주요 변경 사항:

아키텍처 변경:

컨트롤러의 책임을 URL 매핑과 입력 검증으로 한정
서비스는 오로지 도메인 로직에만 집중(서비스와 Request DTO와의 결합 제거)
개별 유스케이스들을 총괄하는 ProductApplicationService 도입

컨트롤러는 이제 ProductApplicationService에만 의존하여 lean해짐

도메인 모델과 DTO 간 변환을 담당하는 ProductMapper를 빈(Bean)으로 전환하여 재사용성 향상

DTO 및 도메인 모델:

불변성과 간결성을 위해 ProductRequest, ProductResponse를 클래스에서 record 타입으로 마이그레이션

Product 모델 생성 로직을 캡슐화하기 위해 팩토리 메서드(of) 패턴 적용

테스트:

Deprecated로 사라질 예정인 MockBean을 제거.

통합 테스트(ProductTest)에 BDD(Given-When-Then) 스타일과 AAA 패턴 적용하여 테스트 목적과 시나리오 가독성 향상

Hyeonjun0527 committed:
refactor(product): 헥사고날 아키텍처 적용 및 기존 모듈 제거

- 헥사고날 아키텍처
spring-gift-wishlist/
└── src/
    └── main/
        └── java/
            └── gift/
                ├── common/                (공통 코드: 예외, 어노테이션, 유틸 등)
                └── product/
                    ├── adapter/           (인프라 계층: 외부와 연결)
                    │   ├── web/           (Primary Adapter: Controller 등)
                    │   └── persistence/   (Secondary Adapter: Repository 구현 등)
                    │
                    ├── application/       (애플리케이션 계층: 비즈니스 워크플로우)
                    │   ├── port/          (포트: 계층 간 추상화 인터페이스)
                    │   │   ├── in/        (Input Port: Use Case 인터페이스)
                    │   │   └── out/       (Output Port: 외부 의존 인터페이스)
                    │   └── usecase/       (Use Case 구현체: 실제 워크플로우/조정)
                    │
                    └── domain/            (도메인 계층: 핵심 비즈니스 로직, 엔티티, 도메인 서비스)

- 기존 애플리케이션 서비스 계층 삭제
- 포트와 어댑터 기반 아키텍처로 리팩터링: AddProductUseCase, DeleteProductUseCase 등 분리
- @Adapter와 UseCase 어노테이션 추가로 어댑터 역할 명시
- SOLID의 ISP 원칙에 따라 컨트롤러가 각 유스케이스 인터페이스를 개별 주입받도록 변경

Hyeonjun0527 committed:
feat(readme): 학습계획 생성

Hyeonjun0527 committed:
feat(MemberPersistenceAdapter): Member Secondary Port 계층 구현
ProductPersistenceAdapter 계층 리팩토링

Hyeonjun0527 committed:
feat(authentication): JWT 기반 인증 및 회원가입/로그인 기능 추가

- JwtTokenPort 및 JwtTokenProvider 구현으로 JWT 생성, 검증, 정보 추출 기능 지원
- SimplePasswordEncoder
- LoginMemberUseCase, SignUpMemberUseCase 인터페이스
- MemberInteractor에서 회원가입 및 로그인 로직 구현
  - Access Token 및 Refresh Token 발급 기능
  - Refresh Token을 이용한 Access Token 갱신
- 회원 로그인 및 회원가입 요청/응답 DTO (LoginRequest, SignUpRequest, AuthResponse)
- 이전커밋에서 누락된 Member

Hyeonjun0527 committed:
feat(jwt, member): JWT 기반 인증 필터 추가 및 역할(Role) 기반 회원 관리 도입

- JwtAuthenticationFilter JWT 요청 필터링 기능
- MEMBER 도메인에 Role(enum) 추가로 USER와 ADMIN 역할 구분
- JwtTokenProvider에서 토큰 생성 시 Role 포함
- AuthResponse DTO 단순화 및 refresh token 관련 로직 제거

Hyeonjun0527 committed:
feat(jwt, member): JWT 기반 인증 필터 추가 및 역할(Role) 기반 회원 관리 도입

- JwtAuthenticationFilter JWT 요청 필터링 기능
- MEMBER 도메인에 Role(enum) 추가로 USER와 ADMIN 역할 구분
- JwtTokenProvider에서 토큰 생성 시 Role 포함
- AuthResponse DTO 단순화 및 refresh token 관련 로직 제거

Hyeonjun0527 committed:
refactor(member): SignUp을 Register로 이름 변경

Hyeonjun0527 committed:
refactor(member): SignUp을 Register로 이름 변경

Hyeonjun0527 committed:
feat(admin): 관리자 회원 관리 및 인증 기능 추가

- AdminController
- @RequireAdmin 어노테이션 추가
  - 관리자 권한이 필요한 메서드에 적용
- AuthorizationAspect 추가
  - Controller 호출 전 관리자 권한을 검증하도록 인터셉터 구현
- HTTP 요청 인증 실패 시 UnauthorizedException,권한 부족 시 ForbiddenException 반환
- WebConfig에 AuthorizationAspect 등록
- MemberPersistencePort에 회원 CRUD 관련 메서드 추가

Hyeonjun0527 committed:
refactor(jwt): 테스트 환경에서 MockJwtTokenProvider 추가 및 JWT 관련 설정 개선

- MockJwtTokenProvider 구현 및 테스트 프로파일 추가
  - 테스트 환경에서 JWT 토큰 생성을 위한 Mock 객체
- JwtTokenProvider 및 관련 빈들에 ConditionalOnProperty 및 Profile 설정
- WebConfig 테스트 프로파일
- Member domain에 Role 추가 및 관련 메서드 수정
  - Member 엔티티 및 관련 repository 메서드 리팩터링
  - schema.sql에 회원 테이블에 role 컬럼 추가 (기본값 USER)
- 테스트 환경 설정을 위해 각종 테스트 클래스에 ActiveProfiles("test") 추가
- JwtTokenPort 인터페이스 메서드 정리
- application-test.properties 추가 및 테스트 환경에서 jwt.enabled 설정 끔

Hyeonjun0527 committed:
feat(exception): 인증 및 권한 오류 처리 기능 강화

- ErrorCode에 인증 및 권한 관련 HTTP 상태 코드 추가
  - UNAUTHORIZED: 인증 자격 증명 필요
  - FORBIDDEN: 접근 권한 없음
- GlobalExceptionHandler UnauthorizedException, ForbiddenException
  - ErrorResponse 대신 ProblemDetail로 통일된 형식 제공
- NumberFormatException 처리 추가

Hyeonjun0527 committed:
docs(develop-diary): docs 작성

Hyeonjun0527 committed:
docs(develop-diary): docs 작성

Hyeonjun0527 committed:
fix(persistence):
- ProductPersistenceAdapter와 MemberPersistenceAdapter에서 기존 getKey() 메서드에서 getKeys 메서드에서 () 명시적으로 ID 키를 가져오도록 수정.
Member에 createdAt 때문에 keyHolder에 두가지 Key가 들어가기 때문
이메일 검증시 잘못된 이메일도 받았음. @Valid 어노테이션 빼먹어서 추가

Hyeonjun0527 committed:
fix(persistence):
getFieldErrors().stream().map으로 errors리스트에 프라블럼 디테일을 리턴하여 에러 객체가 중복 정보를 담게 되는 문제 해결
{
  "type": "about:blank",
  "title": "Bad Request",
  "status": 400,
  "detail": "입력하신 정보를 다시 확인해주세요",
  "instance": "/api/members/register",
  "code": "VALIDATION_FAILED",
  "errors": [
    {
      "type": "about:blank",
      "title": "Bad Request",
      "status": 400,
      "detail": "입력하신 정보를 다시 확인해주세요",
      "code": "VALIDATION_FAILED",
      "field": "email",
      "rejectedValue": "invalid-email1"
    }
  ],
  "path": "/api/members/register"
}

Hyeonjun0527 committed:
refactor(core): 유스케이스 인터페이스 각각 MemberUseCase와 ProductUseCase로 통합
너무 생산성이 떨어지고 불편한 거 같아 제거 단일책임원칙을 지키는 것보다 내려놓는 게 좋다고 생각함

Hyeonjun0527 committed:
fix(jwt): JWT 인증 실패 시 Unauthorized 응답 처리 추가
- 인증 실패 시 ProblemDetail을 JSON 형식으로 반환하도록 수정
- 인증 제외 경로 `/api/members/register`, `/api/members/login`

Hyeonjun0527 committed:
refactor(security): 쓸모없는 throws 제거

jaenyeong commented on DEVELOP-DIARY.md:
```diff
@@ -93,8 +93,103 @@ DDD/CA(도메인 드라이븐 디자인, 클린아키텍처)에서는 도메인
+  2. `@MockitoBean` 사용 → 매 테스트마다 추가 필요로 불편
+  3. 최종 채택: 스프링 프로필 기반 조건부 빈 등록
+- 해결책: 
+  - 프로덕션: `@Profile("!test")` + `@ConditionalOnProperty(jwt.enabled=true)`로 실제 JWT 컴포넌트 로드
```
🟢 
가능하면 낫을 붙이기보단 적용되어야 할 프로파일을 모두 선언하시는 것을 추천 드려요!
이런 경우 test를 제외한 모든 프로파일을 명시할 때 불편하지 않나 라는 생각을 하실 수도 있는데요. 문제는 중간에 다른 페이즈 환경이 추가되는 경우 기능과 설정이 의도하지 않게 적용될 수 있어요 ㅎㅎ

Hyeonjun0527 replied:
넵 조언 감사드립니다!

jaenyeong commented on src/main/java/gift/common/config/WebConfig.java:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    @Configuration
+    @Profile("test")
+    public static class TestWebConfig implements WebMvcConfigurer {
+    }
```
🟢 
크리티컬 한 건 아니지만 테스트를 위한 설정은 테스트 경로에 위치하는게 어떨까요?
성향 차이긴 한데 테스트가 프로덕션을 오염시키는 것을 별로 좋아하진 않아서요 ㅎㅎ

Hyeonjun0527 replied:
아 넵 테스트 경로에 위치해도 되는지를 생각을 안해봤었습니다 찾아보고 적용할 수 있으면 적용하도록 하겠습니다!

jaenyeong commented on src/main/java/gift/common/config/WebConfig.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Configuration
+    @Profile("!test")
+    public static class SecurityWebConfig implements WebMvcConfigurer {
```
❓ 
이너 클래스로 선언하신 건 어떤 의미가 있는 걸까요?
이를 위해 `@Configuartion` 애너테이션을 두번씩 붙이신 것 같아서 여쭤봅니다~

jaenyeong commented on src/main/java/gift/common/exception/GlobalExceptionHandler.java:
```diff
@@ -22,19 +20,37 @@
 
 @RestControllerAdvice
-@Order
+@Order(1)
```
❓ 
저번에도 여쭤본 것 같은데요, 이걸 어떤 의미로 붙이셨을까요?

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,116 @@
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+
+    private static final Logger log = LoggerFactory.getLogger(JwtAuthenticationFilter.class);
+    private static final List<String> SKIP_PATHS = Arrays.asList(
```
🔴 
register(signup), login, root(/) 패스들도 토큰 확인을 스킵해야 하지 않을까요?

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,116 @@
+        try {
+            String token = jwtTokenPort.resolveToken(request);
+            
+            if (token != null) {
```
🟢 
```java
if (token == null) {
    ....
    return;
}
```
처럼 걸었다면 아래 라인의 인덴트를 줄일 수 있지 않았을까요? ㅎㅎ

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,116 @@
+            if (token != null) {
+                JwtTokenProvider.TokenValidationResult validationResult = jwtTokenPort.validateToken(token);
+                
+                if (validationResult.isValid()) {
```
🟢 
마찬가지로
```java
if (validationResult.isNotValid()) {
    ...
    return;
}
```
였다면 어떨까요?

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,116 @@
+                    
+                    log.info("JWT 인증 성공: {} (ID: {}), role: {}", email, memberId, role);
+                } else {
+                    log.info("유효하지 않은 JWT 토큰: {}", validationResult.getErrorMessage());
```
🟢 
이곳을 포함해 정상(플로우) 동작이 아니라고 생각되는 경우엔 로그가 info 레벨이 아니라 warn 또는 error 레벨로 찍혀야 하지 않을까요?

jaenyeong commented on src/test/java/gift/common/exception/GlobalExceptionHandlerTest.java:
```diff
@@ -19,6 +20,7 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
 
 @WebMvcTest(ProductController.class)
+@ActiveProfiles("test")
```
🔴 
해당 테스트 클래스는 빌드해보지 않으신 것 같네요 ㅎㅎ 예전 클래스들을 의존해서 컴파일 에러가 나네요~

jaenyeong commented on src/main/java/gift/common/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,116 @@
+@ConditionalOnProperty(
+    name = "jwt.enabled", 
+    havingValue = "true", 
+    matchIfMissing = true
```
❓ 
해당 속성이 true인 이유는 무엇일까요?

jaenyeong reviewed (changes requested):
현준님~ 고생 하셨습니다~
조금 확인해 주셔야 하는 부분을 코멘트 드렸구요!
🟢 코멘트는 참고만 해주세요~
전반적으로 파일 끝에 trailing new line을 없애신 것 같네요? ㅎㅎ
이건 아마 당장 큰 문제는 없지만 posix 계열 표준을 따르는 유틸 등에서 문제가 발생할 순 있습니다~

너무 많은 코멘트를 남기는 건 조금 지양하려고 해서, 중점 리뷰 부분에 대해서는 간략히 말씀만 드릴테니 추가 질문이 있다면 코멘트로 직접 남겨주시면 답변 드릴게요 ㅎㅎ

> JwtAuthenticationFilter으로 관리자용 api를 보호

관리자용 이라는게 결국 Role에 따라 인가를 말씀하시는 것 같은데 위 필터는 이름만으로 봤을때 인증 필터기 때문에 인가 필터를 따로 추가하는 것이 어떨까 하는 의견을 드려봅니다~ ㅎㅎ

> JwtTokenProvider로 토큰 생성/검증/추출 하는 부분

이 부분은 스타일에 따라 리팩터링이 될 순 있겠으나 크게 중요하진 않을 것 같아서 뺐습니다~
원하시는 질문이 있으면 따로 남겨주세요!

> 컨트롤러가 각 모든 UseCase를 의존하는데 파사드 패턴으로 모든 Product 관련 UseCase를 묶어 Facade를 만들지 고민입니다.

지금은 간단해서 그렇게 해도 괜찮을 것 같은데 원하시는 퍼사드의 형태를 직접 봐야 코멘트를 드릴 수 있을 것 같아요~

> 그리고 UseCase라는 인터페이스명을 Port로 변경할 지 고민입니다.

이 부분은 결정하기 나름이라고 생각되는데 개인적으로 포트는 결국 인터페이스를 표현하니 구현체에서는 굳이 라는 생각이 있긴 하네요~ 다만 소속된 팀의 협의에 따라 다를 것 같아요~

아무튼 수정 한 번 부탁드려요~

Hyeonjun0527 committed:
refactor(all):리뷰 반영 - 테스트 환경 bean 분리
  - TestConfig, MockJwtTokenProvider
  MockJwtTokenProvider는 MemberInteractor에서 사용중이기에 Mock가 필요했음.
  - admin.enabled, jwt.enabled 설정 값 조건부 로딩 추가
  - 현재 상황에 불필요한 @Order 제거
- ProductTest, ProductControllerTest에서 @ActiveProfiles 제거
- JwtTokenProvider에서 isValid 메서드를 isNotValid로 변경
- JwtAuthenticationFilter 가독성 향상
- WebConfig와 AuthorizationAspect 등에서 혼란을 야기하는 profile 기반 설정 +  ConditionalOnProperty에서 오직 ConditionalOnProperty로만 관리

jaenyeong reviewed (approved):
현준님 고생하셨습니다~
일단 다음 스텝으로 넘어가시고, 제가 질문 드린 것들은 현준님의 의견에 따라 제가 코멘트가 바뀔 수 있는 부분이라 다음 스텝 진행할 때 고민해 보셨으면 합니다 :)
머지할게요 ㅎㅎ
